vortex-serde/src/layouts/reader/mod.rs

```diff
@@ -1,12 +1,123 @@
+use std::fmt::Debug;
+use std::sync::{Arc, RwLock};
+
+use ahash::HashMap;
+use bytes::Bytes;
+pub use layouts::{ChunkedLayoutSpec, ColumnLayoutSpec};
 use projections::Projection;
+use vortex::Array;
+use vortex_dtype::DType;
+use vortex_error::VortexResult;
+
+use crate::layouts::reader::filtering::RowFilter;
+use crate::writer::ByteRange;
 
 pub mod batch;
 pub mod buffered;
 pub mod builder;
+pub mod context;
 pub mod filtering;
+mod footer;
+mod layouts;
 pub mod projections;
 pub mod schema;
 pub mod stream;
 
 const DEFAULT_BATCH_SIZE: usize = 65536;
 const DEFAULT_PROJECTION: Projection = Projection::All;
+
+#[derive(Debug, Clone)]
+pub struct Scan {
+    indices: Option<Array>,
+    projection: Projection,
+    filter: Option<RowFilter>,
+    batch_size: usize,
+}
+
+/// Unique identifier for a message within a layout
+pub type LayoutPartId = u16;
+pub type MessageId = Vec<LayoutPartId>;
+
+#[derive(Debug)]
+pub enum ReadResult {
+    GetMsgs(Vec<(MessageId, ByteRange)>),
+    Batch(Array),
+}
+
+#[derive(Default, Debug)]
+pub struct LayoutMessageCache {
+    cache: HashMap<MessageId, Bytes>,
+}
+
+impl LayoutMessageCache {
+    pub fn get(&self, path: &[LayoutPartId]) -> Option<Bytes> {
+        self.cache.get(path).cloned()
+    }
+
+    pub fn remove(&mut self, path: &[LayoutPartId]) -> Option<Bytes> {
+        self.cache.remove(path)
+    }
+
+    pub fn set(&mut self, path: MessageId, value: Bytes) {
+        self.cache.insert(path, value);
+    }
+}
+
+#[derive(Debug)]
+pub struct RelativeLayoutCache {
+    parent: Arc<RwLock<LayoutMessageCache>>,
+    path: MessageId,
+    dtype: DType,
+}
+
+impl RelativeLayoutCache {
+    pub fn new(dtype: DType, parent: Arc<RwLock<LayoutMessageCache>>, path: MessageId) -> Self {
```

Comment:
wdyt about having two constructs:
1. Externally facing one that doesn't take a path
2. Internal `with_path` to be used in `relative`

Reply:
I just construct the struct directly